## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [35]:
import pandas as pd



In [36]:
df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [37]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [38]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [39]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# Display data types 
print(df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [40]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# Display missing values
print(df.isnull().sum())

first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64


In [41]:
# What three item categories had the most entries?
category_counts = df['category'].value_counts()
top_three_categories = category_counts.head(3)
top_three_categories

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [52]:
# For the category with the most entries, which subcategory had the most entries?
category_counts_df = df['category'].value_counts()
most_common_category = category_counts_df.idxmax()

# Filter the dataset for the most common category
most_common_category_data = df[df['category'] == most_common_category]
consumable_subcategory_counts = consumable_data['subcategory'].value_counts()

# Determine the subcategory with the most entries within the most common category
subcategory_counts = most_common_category_data['subcategory'].value_counts()
most_common_subcategory = subcategory_counts.idxmax()

print(f"The category with the most entries is: {most_common_category}")
print(f"The subcategory with the most entries within {most_common_category} is: {most_common_subcategory}")



The category with the most entries is: consumables
The subcategory with the most entries within consumables is: bathroom supplies


In [56]:
# Which five clients had the most entries in the data?

client_counts = df['client_id'].value_counts()
top_five_clients = client_counts.head(5)

print("The five clients with the most entries are:")
print(top_five_clients)

The five clients with the most entries are:
client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [57]:
# Store the client ids of those top 5 clients in a list.
top_five_client_ids = top_five_clients.index.tolist()

print("The client IDs of the top 5 clients are:")
print(top_five_client_ids)


The client IDs of the top 5 clients are:
[33615, 66037, 46820, 38378, 24741]


In [58]:
# How many total units (the qty column) did the client with the most entries order?

# top_client_id is 33615
top_client_id = df['client_id'].value_counts().idxmax()
top_client_id

# Filter the dataset for this client ID
top_client_id_data = df[df['client_id'] == top_client_id]
top_client_id_data

# Sum the qty column for this client
total_units_ordered = top_client_id_data['qty'].sum()

print(f'Total units ordered by the client with the most entries (ID {top_client_id}): {total_units_ordered}')

Total units ordered by the client with the most entries (ID 33615): 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [59]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty

df['subtotal'] = df['unit_price'] * df['qty']
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [60]:
# Create a column for shipping price. Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

# Calculate the total weight for each order line
df['total_weight'] = df['unit_weight'] * df['qty']
# df.head()

# Create a new column 'shipping_price' based on the total weight
df['shipping_price'] = df['total_weight'].apply(lambda x: 7 * x if x > 50 else 10 * x)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  total_weight  \
0     762.71         7.50  105            1  115164.00        787.50   
1      15.09         1.49   21            0     523.95         31.29   
2       7.86         1.68   39            6     527.28         65.52   
3      24.85         1.23   29            3    1056.18         35.67   
4     108.17        46.43   20            1    3902.00        928.60   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 21 columns]

In [61]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

# Create a new column 'subtotal' by multiplying 'unit_price' and 'qty'
df['subtotal'] = df['unit_price'] * df['qty']

# Calculate the total price including sales tax
df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925

# Display the first few rows to verify the new columns
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  total_weight  shipping_price  \
0         7.50  105            1  115164.00        787.50         5512.50   
1         1.49   21            0     523.95         31.29          312.90   
2         1.68   39            6     527.28         65.52          458.64   
3         1.23   29            3    1056.18         35.67          356.70   
4        46.43   20            1    3902.00        928.60         6500.20   

     total_price  
0  131839.076250  
1     914.258625  
2    1077.117600  
3    1543.571400  
4   11364.403500  

[5 rows x 22 columns]

In [62]:
# Create a column for the cost of each line using unit cost, qty, and shipping price (assume the shipping cost is exactly what is charged to the client).

# Calculate the cost for each line
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  total_weight  shipping_price    total_price  \
0            1  115164.00        787.50         5512.50  131839.076250   
1            0     523.95         31.29          312.90     914.258625   
2            6     527.28         65.52          458.64    1077.117600   
3            3    1056.18         35.67          356.70    1543.571400   
4            1    3902.00        928.60         6500.20   11364.403500   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  
3    1077.35  
4    8663.60  

[5 rows x 23 columns]

In [63]:
# Create a column for the profit of each line using line cost and line price

# Calculate the profit for each line
df['profit'] = df['total_price'] - df['line_cost']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    subtotal  total_weight  shipping_price    total_price  line_cost  \
0  115164.00        787.50         5512.50  131839.076250   85597.05   
1     523.95         31.29          312.90     914.258625     629.79   
2     527.28         65.52          458.64    1077.117600     765.18   
3    1056.18         35.67          356.70    1543.571400    1077.35   
4    3902.00        928.60         6500.20   11364.403500    8663.60   

         profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [64]:
# Check your work using the totals above

# For Order ID 2742071 with a total price of $152,811.89
order_data = df[df['order_id'] == 2742071]
order_data 

total_price_sum = order_data['total_price'].sum()
total_price_sum
print(f'Total price for order ID 2742071: ${total_price_sum:.2f}')


Total price for order ID 2742071: $152811.90


In [65]:
# Order ID 2173913 had a total price of $162,388.71
order_data = df[df['order_id'] == 2173913]
order_data 

total_price_sum = order_data['total_price'].sum()
total_price_sum
print(f'Total price for order ID 2173913: ${total_price_sum:.2f}')


Total price for order ID 2173913: $162388.72


In [66]:
# Order ID 6128929 had a total price of $923,441.25
order_data = df[df['order_id'] == 6128929]
order_data 

total_price_sum = order_data['total_price'].sum()
total_price_sum
print(f'Total price for order ID 6128929: ${total_price_sum:.2f}')

Total price for order ID 6128929: $923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [68]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
# define a function that will revieve the client id and for loop grap the line price

top_5_data = df[df['client_id'].isin(top_five_client_ids)]
top_5_data


first     last                                           job  \
5        Jessica    Reyes  Production designer, theatre/television/film   
12       Jessica    Reyes  Production designer, theatre/television/film   
137        Bryan    Myers                                Chief of Staff   
214        Bryan    Myers                                Chief of Staff   
216       Angela  Everett                                     Osteopath   
...          ...      ...                                           ...   
54535     Angela  Everett                                     Osteopath   
54536     Kendra  Garrett                        Pension scheme manager   
54579  Alexandra    Young                                     Dietitian   
54589  Alexandra    Young                                     Dietitian   
54613     Kendra  Garrett                        Pension scheme manager   

                  phone                            email  client_id  order_id  \
5            5016915923         jessica.reyes@grimes.net      33615   1327730   
12           5016915923         jessica.reyes@grimes.net      33615   6059361   
137    444.233.9146x233            bmyers3696@wilson.com      46820   7109502   
214    444.233.9146x233            bmyers3696@wilson.com      46820   8967498   
216    729-620-6192x875           aeverett@rodriguez.com      66037   1203693   
...                 ...                              ...        ...       ...   
54535  729-620-6192x875           aeverett@rodriguez.com      66037   7550600   
54536      808.257.2029  garrettkendra8479@jefferson.com      24741   1886506   
54579  265-666-3155x118           ayoung9158@stevens.com      38378   8661147   
54589  265-666-3155x118           ayoung9158@stevens.com      38378   4508675   
54613      808.257.2029  garrettkendra8479@jefferson.com      24741   9114777   

       order_date  order_week  order_year  ... unit_cost unit_weight  qty  \
5      2023-05-02          18        2023  ...    154.44        2.79   78   
12     2023-05-16          20        2023  ...    198.38       27.51  568   
137    2023-05-12          19        2023  ...    111.47        5.65  118   
214    2023-02-06           6        2023  ...      6.26        7.05  133   
216    2023-04-09          14        2023  ...    166.09        0.26  157   
...           ...         ...         ...  ...       ...         ...  ...   
54535  2023-02-24           8        2023  ...     10.66        7.52  108   
54536  2023-05-26          21        2023  ...     40.05        4.70  721   
54579  2023-03-09          10        2023  ...     20.29       13.07  118   
54589  2023-03-25          12        2023  ...     38.26        0.70  685   
54613  2023-02-13           7        2023  ...     59.32       13.13   91   

       line_number   subtotal  total_weight  shipping_price    total_price  \
5                0   13481.52        217.62         1523.34   16392.809550   
12               2  130492.32      15625.68       109379.76  262060.247400   
137              9   16063.34        666.70         4666.90   22647.787200   
214              3    1061.34        937.65         6563.55    8330.192325   
216              2   47393.59         40.82          408.20   52223.455575   
...            ...        ...           ...             ...            ...   
54535            2    1810.08        812.16         5685.12    8188.506000   
54536            3   46014.22       3388.70        23720.90   76185.618600   
54579            3    3687.50       1542.26        10795.82   15823.027100   
54589            0   35784.40        479.50         3356.50   42761.433250   
54613            2    8705.97       1194.83         8363.81   18648.734650   

       line_cost        profit  
5       13569.66   2823.149550  
12     222059.60  40000.647400  
137     17820.36   4827.427200  
214      7396.13    934.062325  
216     26484.33  25739.125575  
...          ...           ...  
54535    6836.40   1352.106000  
54536   525

In [72]:
## Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
## total units purchased, total shipping price, total revenue, and total profit. 

client_counts = df['client_id'].value_counts()
top_five_clients = client_counts.head(5)
top_five_client_ids = top_five_clients.index.tolist()

top_5_data = df[df['client_id'].isin(top_five_client_ids)]



# Calculate the required metrics for the top 5 clients
# define a function and for loop
summary = top_5_data.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('total_price', 'sum'),
    total_profit=('profit', 'sum')
).reset_index()

print(summary)

   client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      24741                 239862            5126448.37   8.226889e+07   
1      33615                  64313            1828984.89   8.377309e+06   
2      38378                  73667            3429455.40   1.290655e+07   
3      46820                  75768            1601448.84   9.743794e+06   
4      66037                  43018            1395151.85   1.025951e+07   

   total_profit  
0  3.657999e+07  
1  2.201995e+06  
2  3.271830e+06  
3  2.736603e+06  
4  3.255032e+06  


In [73]:
# Format the data and rename the columns to names suitable for presentation.

display_summary_df = summary.rename(columns={
    'client_id': 'Client ID',
    'total_units_purchase': 'Units',
    'total_shipping_price': 'Shipping Price',
    'total_revenue': 'Total Revenue',
    'total_profit': 'Total Profit'
})

print(display_summary_df)


# Define a function that converts a dollar amount to millions.
def scale_to_millions(amount):
    return amount / 1000000


   Client ID  total_units_purchased  Shipping Price  Total Revenue  \
0      24741                 239862      5126448.37   8.226889e+07   
1      33615                  64313      1828984.89   8.377309e+06   
2      38378                  73667      3429455.40   1.290655e+07   
3      46820                  75768      1601448.84   9.743794e+06   
4      66037                  43018      1395151.85   1.025951e+07   

   Total Profit  
0  3.657999e+07  
1  2.201995e+06  
2  3.271830e+06  
3  2.736603e+06  
4  3.255032e+06  


In [74]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.

sorted_data = display_summary_df.sort_values(by='Total Profit', ascending=False)
sorted_data

Client ID  total_units_purchased  Shipping Price  Total Revenue  \
0      24741                 239862      5126448.37   8.226889e+07   
2      38378                  73667      3429455.40   1.290655e+07   
4      66037                  43018      1395151.85   1.025951e+07   
3      46820                  75768      1601448.84   9.743794e+06   
1      33615                  64313      1828984.89   8.377309e+06   

   Total Profit  
0  3.657999e+07  
2  3.271830e+06  
4  3.255032e+06  
3  2.736603e+06  
1  2.201995e+06